File: src/main/java/io/github/drmanan/learn/sort/InsertionSort.java
/**
 * <H2> Insertion </H2>
 *
 * @author Manan Sharma
 * @version 1
 * @since Wednesday, 16 of August, 2023; 17:51:56
 */

package io.github.drmanan.learn.sort;

import io.github.drmanan.learn.commons.Array;

import static io.github.drmanan.learn.commons.Array.POSITION.AFTER;
import static io.github.drmanan.learn.commons.Array.POSITION.BEFORE;
import static io.github.drmanan.learn.commons.Array.generate_random_array;
import static io.github.drmanan.learn.commons.Array.print_array;

public class InsertionSort {

    // Driver method
    public static void main(String[] args) {
        int[] arr = generate_random_array(20, 100);
        // int[] arr = {12, 11, 13, 5, 6};
        // int[] arr = {4, 3, 5, 2, 1};
        print_array(BEFORE, arr);
        InsertionSort sort = new InsertionSort();
        sort.sort(arr);
        print_array(AFTER, arr);
    }

    /*Function to sort the array using insertion sort*/
    void sort(int[] arr) {
        for (int i = 1; i < arr.length; ++i) {
            int key = arr[i];
            int j = i - 1;
            print_array((i - 1), arr, j, i);
            /* Move elements of arr[0...i-1], that are
            greater than the key, to one position ahead
            of their current position */
            while (j >= 0 && arr[j] > key) {
                arr[j + 1] = arr[j];
                j = j - 1;
            }
            arr[j + 1] = key;
            print_array((i - 1), arr, j+1, i);
        }
    }
}

File: src/main/python/sort/InsertionSort.py
from src.main.python.commons.Array import generate_random_array
from src.main.python.commons.Array import print_array


def insertion_sort(arrar_to_sort):
    # Traverse through 1 to len(arr)
    for i in range(1, len(arrar_to_sort)):
        key = arrar_to_sort[i]
        # Move elements of arr[0..i-1], that are
        # greater than key, to one position ahead
        # of their current position
        j = i - 1
        while j >= 0 and key < arrar_to_sort[j]:
            arrar_to_sort[j + 1] = arrar_to_sort[j]
            j -= 1
        print_array(arrar_to_sort, step_number=i)
        arrar_to_sort[j + 1] = key  # Driver code to test above


array = generate_random_array(10, 10)
print_array(array, position="BEFORE")
insertion_sort(array)
print_array(array, position="AFTER")
